plonky2/src/hash/hashing.rs

```diff
@@ -90,14 +92,22 @@ pub trait PlonkyPermutation<F: RichField> {
 /// for the hash to be secure, but it can safely be disabled in certain cases, like if the input
 /// length is fixed.
 pub fn hash_n_to_m_no_pad<F: RichField, P: PlonkyPermutation<F>>(
-    inputs: &[F],
+    inputs: impl IntoIterator<Item = impl Borrow<F>>,
     num_outputs: usize,
 ) -> Vec<F> {
     let mut state = [F::ZERO; SPONGE_WIDTH];
 
     // Absorb all input chunks.
-    for input_chunk in inputs.chunks(SPONGE_RATE) {
-        state[..input_chunk.len()].copy_from_slice(input_chunk);
+    let mut pos = 0;
+    for input in inputs.into_iter() {
+        if pos == SPONGE_RATE {
+            state = P::permute(state);
+            pos = 0;
+        }
+        state[pos] = *input.borrow();
+        pos += 1;
+    }
+    if pos != 0 {
         state = P::permute(state);
     }
```

Comment:
```suggestion
    let inputs = inputs.into_inter().fused();
    while Some(state_0) = inputs.next() {
        state[0] = *state_0.borrow();
        for (state_i, i) in inputs.zip(1..SPONGE_RATE) {
            state[i] = *state_i.borrow();
        }
        state = P::permute(state);
    }
```

Reply:
Good idea; changed. I'm getting similar performance, but it is nicer at least.

I just had to switch the `zip` order, to avoid zip consuming an item from `inputs` before realizing the other iterator was exhausted. Took me a while to notice that 😅

Reply:
(Though take my perf claims with a grain of salt, haven't done proper side by side comparisons)

Reply:
> consuming an item from `inputs` before realizing the other iterator was exhausted

Oh true! oops